objectstore-service/src/backend/s3_compatible.rs

```
@@ -1,131 +1,117 @@
+            && path_style
+        {
+            bucket = bucket.with_path_style();
         }
-    }
 
-    /// Formats the S3 object URL for the given key.
-    fn object_url(&self, path: &ObjectPath) -> String {
-        format!("{}/{}/{path}", self.endpoint, self.bucket)
-    }
-}
-
-impl<T> S3CompatibleBackend<T>
-where
-    T: TokenProvider,
-{
-    /// Creates a request builder with the appropriate authentication.
-    async fn request(&self, method: Method, url: impl IntoUrl) -> Result<RequestBuilder> {
-        let mut builder = self.client.request(method, url);
-        if let Some(provider) = &self.token_provider {
-            builder = builder.bearer_auth(provider.get_token().await?.as_str());
+        if let Some(request_timeout) = config.request_timeout {
+            bucket = bucket.with_request_timeout(request_timeout).unwrap();
         }
-        Ok(builder)
-    }
-
-    /// Issues a request to update the metadata for the given object.
-    async fn update_metadata(&self, path: &ObjectPath, metadata: &Metadata) -> Result<()> {
-        // NB: Meta updates require copy + REPLACE along with *all* metadata. See
-        // https://cloud.google.com/storage/docs/xml-api/put-object-copy
-        self.request(Method::PUT, self.object_url(path))
-            .await?
-            .header("x-goog-copy-source", format!("/{}/{path}", self.bucket))
-            .header("x-goog-metadata-directive", "REPLACE")
-            .headers(metadata.to_headers(GCS_CUSTOM_PREFIX, true)?)
-            .send()
-            .await?
-            .error_for_status()
-            .context("failed to update expiration time for object with TTI")?;
 
-        Ok(())
+        Self { bucket }
+    }
+
+    async fn handle_get_object_response(
+        &self,
+        _path: &ObjectPath,
+        response: ResponseData,
+    ) -> Result<Option<(Metadata, BackendStream)>> {
+        let metadata = Metadata::from_hashmap(response.headers(), "").unwrap_or_default();
```

this should propagate any error that happens.